nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java

```diff
@@ -150,548 +145,320 @@
-                try (final BufferedInputStream bufferedIn = new BufferedInputStream(rawIn);
-                        final ByteCountingInputStream in = new ByteCountingInputStream(bufferedIn)) {
-
-                    long bufferedPartialLine = 0;
-
-                    // if we have header lines, copy them into a ByteArrayOutputStream
-                    final ByteArrayOutputStream headerStream = new ByteArrayOutputStream();
-                    // Determine the number of lines of header, priority given to HEADER_LINE_COUNT property
-                    int headerInfoLineCount = 0;
-                    if (headerCount > 0) {
-                        headerInfoLineCount = headerCount;
-                    } else {
-                        if (headerMarker != null) {
-                            headerInfoLineCount = countHeaderLines(in, headerMarker);
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile != null) {
+            AtomicBoolean error = new AtomicBoolean();
+            List<FlowFile> splitFlowFiles = new ArrayList<>();
+            List<SplitInfo> computedSplitsInfo = new ArrayList<>();
+            AtomicReference<SplitInfo> headerSplitInfoRef = new AtomicReference<>();
+            session.read(flowFile, new InputStreamCallback() {
+                @Override
+                public void process(InputStream in) throws IOException {
+                    TextLineDemarcator demarcator = new TextLineDemarcator(in);
+                    SplitInfo splitInfo = null;
+                    long startOffset = 0;
+
+                    // Compute fragment representing the header (if available)
+                    long start = System.nanoTime();
+                    try {
+                        if (SplitText.this.headerLineCount > 0) {
+                            splitInfo = SplitText.this.computeHeader(demarcator, startOffset, SplitText.this.headerLineCount, null, null);
+                            if (splitInfo.lineCount < SplitText.this.headerLineCount) {
+                                error.set(true);
+                                getLogger().error("Unable to split " + flowFile + " due to insufficient amount of header lines. Required "
+                                        + SplitText.this.headerLineCount + " but was " + splitInfo.lineCount + ". Routing to failure.");
+                            }
+                        } else if (SplitText.this.headerMarker != null) {
+                            splitInfo = SplitText.this.computeHeader(demarcator, startOffset, Long.MAX_VALUE, SplitText.this.headerMarker.getBytes(StandardCharsets.UTF_8), null);
                         }
+                        headerSplitInfoRef.set(splitInfo);
+                    } catch (IllegalStateException e) {
+                        error.set(true);
+                        getLogger().error(e.getMessage() + " Routing to failure.");
```

Comment:
We should probably be logging e.toString() instead of e.getMessage(), as it provides more details about what went wrong. We should also be logging a toString() of the FlowFile that is being routed to failure.